src/pkg/audit/dao/dao.go

```diff
@@ -34,15 +39,85 @@ type DAO interface {
 	Get(ctx context.Context, id int64) (access *model.AuditLog, err error)
 	// Delete the audit log specified by ID
 	Delete(ctx context.Context, id int64) (err error)
+	// Purge the audit log
+	Purge(ctx context.Context, retentionHour int, includeOperations []string, dryRun bool) (int64, error)
 }
 
 // New returns an instance of the default DAO
 func New() DAO {
 	return &dao{}
 }
 
+var allowedMaps = map[string]string{
+	strings.ToLower(rbac.ActionPull.String()):   "",
+	strings.ToLower(rbac.ActionCreate.String()): "",
+	strings.ToLower(rbac.ActionDelete.String()): "",
+}
+
 type dao struct{}
 
+// Purge delete expired audit log
+func (*dao) Purge(ctx context.Context, retentionHour int, includeOperations []string, dryRun bool) (int64, error) {
+	ormer, err := orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	var sql string
+	if dryRun {
+		sql = "SELECT count(1) cnt FROM audit_log WHERE op_time < NOW() - ? * interval '1 hour' "
+	} else {
+		sql = "DELETE FROM audit_log WHERE op_time < NOW() - ? * interval '1 hour' "
+	}
+
+	filterOps := permitOps(includeOperations)
+	sql = sql + "AND lower(operation) IN ('" + strings.Join(filterOps, "','") + "')"
+	log.Infof("the sql is %v", sql)
+
+	if dryRun {
+		var cnt []int64
+		r, err := ormer.Raw(sql, retentionHour).QueryRows(&cnt)
+		if err != nil {
+			log.Errorf("failed to dry run purge audit log, error %v", err)
+			return 0, err
+		}
+		if r != 1 {
+			return 0, fmt.Errorf("failed to purge audit log, no result found")
+		}
+		log.Infof("[DRYRUN]purged %d audit logs in the database", cnt[0])
+		return cnt[0], nil
+	}
+
+	r, err := ormer.Raw(sql, retentionHour).Exec()
+	if err != nil {
+		log.Errorf("failed to purge audit log, error %v", err)
+		return 0, err
+	}
+	delRows, rErr := r.RowsAffected()
+	if rErr != nil {
+		log.Errorf("failed to purge audit log, error %v", rErr)
+		return 0, rErr
+	}
+	log.Infof("purged %d audit logs in the database", delRows)
+
+	return delRows, err
+}
+
+// permitOps filter not allowed operation, if no operation specified, purge pull operation
+func permitOps(includeOperations []string) []string {
+	var filterOps []string
+	for _, ops := range includeOperations {
+		ops := strings.ToLower(ops)
+		if _, exist := allowedMaps[ops]; exist {
+			filterOps = append(filterOps, ops)
+		}
+	}
+	// purge pull operation by default
+	if len(filterOps) == 0 {
```

Comment:
how does we present the default behavior in UI?

Reply:
default is pull operation only in UI